src/plan/transitive_closure.rs

```diff
@@ -71,3 +71,163 @@ impl<'a, E: ProcessEdgesWork> Drop for ObjectsClosure<'a, E> {
+{
+    type VM = VM;
+
+    fn new(edges: Vec<Address>, roots: bool, mmtk: &'static MMTK<VM>) -> Self {
+        let base = ProcessEdgesBase::new(edges, roots, mmtk);
+        let plan = base.plan().downcast_ref::<P>().unwrap();
+        Self { plan, base }
+    }
+
+    #[inline(always)]
+    fn create_scan_work(&self, nodes: Vec<ObjectReference>) -> Box<dyn GCWork<Self::VM>> {
+        Box::new(PlanScanObjects::<Self, P>::new(self.plan, nodes, false))
+    }
+
+    #[inline(always)]
+    fn trace_object(&mut self, object: ObjectReference) -> ObjectReference {
+        if object.is_null() {
+            return object;
+        }
+        self.plan
+            .trace_object::<Self, KIND>(self, object, self.worker())
+    }
+
+    #[inline]
+    fn process_edge(&mut self, slot: Address) {
+        let object = unsafe { slot.load::<ObjectReference>() };
+        let new_object = self.trace_object(object);
+        if P::may_move_objects::<KIND>() {
+            unsafe { slot.store(new_object) };
+        }
+    }
+}
+
+// Impl Deref/DerefMut to ProcessEdgesBase for PlanProcessEdges
+impl<
+        VM: VMBinding,
+        P: 'static + PlanTraceObject<VM> + Plan<VM = VM> + Sync,
+        const KIND: TraceKind,
+    > Deref for PlanProcessEdges<VM, P, KIND>
+{
+    type Target = ProcessEdgesBase<VM>;
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        &self.base
+    }
+}
+
+impl<
+        VM: VMBinding,
+        P: 'static + PlanTraceObject<VM> + Plan<VM = VM> + Sync,
+        const KIND: TraceKind,
+    > DerefMut for PlanProcessEdges<VM, P, KIND>
+{
+    #[inline]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.base
+    }
+}
+
+use std::marker::PhantomData;
+
+/// This provides an implementation of scanning objects work. Each object will be scanned by calling `scan_object()`
+/// in `PlanTraceObject`.
+pub struct PlanScanObjects<
+    E: ProcessEdgesWork,
+    P: 'static + Plan<VM = E::VM> + PlanTraceObject<E::VM> + Sync,
+> {
+    plan: &'static P,
+    buffer: Vec<ObjectReference>,
+    #[allow(dead_code)]
+    concurrent: bool,
+    phantom: PhantomData<E>,
+}
+
+impl<E: ProcessEdgesWork, P: 'static + Plan<VM = E::VM> + PlanTraceObject<E::VM> + Sync>
+    PlanScanObjects<E, P>
+{
+    pub fn new(plan: &'static P, buffer: Vec<ObjectReference>, concurrent: bool) -> Self {
+        Self {
+            plan,
+            buffer,
+            concurrent,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<E: ProcessEdgesWork, P: 'static + Plan<VM = E::VM> + PlanTraceObject<E::VM> + Sync>
+    GCWork<E::VM> for PlanScanObjects<E, P>
+{
+    fn do_work(&mut self, worker: &mut GCWorker<E::VM>, _mmtk: &'static MMTK<E::VM>) {
+        trace!("PlanScanObjects");
+        {
+            let tls = worker.tls;
+            let mut closure = ObjectsClosure::<E>::new(worker);
+            for object in &self.buffer {
+                use crate::vm::Scanning;
```

Comment:
Move this to the top?

Reply:
Do you mean moving this to the top of the file? It is only used in this scope.

Reply:
As Kunshan suggested, this code is moved to `scheduler::gc_work`. We no longer need the `use` in the scope.